### PR TITLE
[Perf] Improve fp8 quant in mla; replace ReduceSum with ReduceScatterSum

### PR DESCRIPTION
**Pull request by IwakuraRein**
## Purpose

- Choose `ReduceScatterSum` over `ReduceSum` when sizes are the same
  
  image

  image

- Avoid calling fp8 quant kernel twice when using fp8 attention by creating a stage buffer for `decode_ql_nope` and `decode_q_pe`. This also eliminates the `torch.cat` in the [flashinfer mla backend](https://github.com/vllm-project/vllm/blob/75957c4e2102a78883d0745962c7986e95316d3b/vllm/v1/attention/backends/mla/flashinfer_mla.py#L135-L137).

  image

## Test Plan

**Launch command**

```
  --dtype auto --kv-cache-dtype fp8 \
  --tensor-parallel-size 8 \
  --swap-space 16 --max-num-seqs 1024 --trust-remote-code --max-model-len 10240 --gpu-memory-utilization 0.95 \
  --max-num-batched-tokens 16384 --async-scheduling \
  --max-cudagraph-capture-size 1024 --compilation_config.cudagraph_mode FULL_DECODE_ONLY
```

**Evaluation command**

```
lm-eval --model local-completions --tasks gsm8k --model_args model=nvidia/DeepSeek-R1-0528-FP4-v2,base_url=http://127.0.0.1:8000/v1/completions,num_concurrent=64,max_retries=3,tokenized_requests=False --num_fewshot 20
```

**Models**

- Test with [DeepSeek R1 0528 checkpoint](https://huggingface.co/deepseek-ai/DeepSeek-R1-0528) and 8 H200.
- Test with [DeepSeek R1 0528 FP4 V2 checkpoint](https://huggingface.co/nvidia/DeepSeek-R1-0528-NVFP4-v2) and 4 GB200.

## Test Result

- Hopper, `VLLM_ATTENTION_BACKEND=FLASHMLA`

  ```
  |Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
  |-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
  |gsm8k|      3|flexible-extract|    20|exact_match|↑  |0.9530|±  |0.0058|
  |     |       |strict-match    |    20|exact_match|↑  |0.9522|±  |0.0059|
  ```

- Blackwell, `VLLM_ATTENTION_BACKEND=FLASHINFER_MLA`

  ```
  |Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
  |-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
  |gsm8k|      3|flexible-extract|    20|exact_match|_  |0.9538|_  |0.0058|
  |     |       |strict-match    |    20|exact_match|_  |0.9522|_  |0.0059|
  ```

Perf gain: 5% projected improvement in decode when using flashinfer mla backend  when serving DeepSeek R1 FP4 on 4 GB200 with DEP4

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/v1/attention/backends/mla/common.py [hunk 1]
```diff
@@ -2028,21 +2028,29 @@ def forward(
...
-                    decode_q_pe.reshape([q_pe_shape[0], q_pe_shape[1] * q_pe_shape[2]]),
-                    layer._q_scale,
+                assert decode_ql_nope.shape[0] == decode_q_pe.shape[0]
+                assert decode_ql_nope.shape[1] == decode_q_pe.shape[1]
+                decode_q_shape = (
+                    ql_nope_shape[0],
+                    ql_nope_shape[1],
+                    ql_nope_shape[2] + q_pe_shape[2],
+                )
+                decode_q0 = torch.empty(
```
Comment (MatthewBonanni) [inline vllm/v1/attention/backends/mla/common.py:2049]: Why not just use `decode_q0 = torch.cat([decode_ql_nope, decode_q_pe], dim=-1)` instead of `torch.empty` + `copy_`?

Also, can we just use `decode_q` instead of the additional intermediate `decode_q0`? i.e.:

```
decode_q = torch.cat([decode_ql_nope, decode_q_pe], dim=-1)
decode_q_shape = decode_q.shape

decode_q, _ = ops.scaled_fp8_quant(
    decode_q.view(decode_q_shape[0], -1),
    layer._q_scale,
)
decode_q = decode_q.view(decode_q_shape)
```

**Review by MatthewBonanni — COMMENTED** [review]
Thanks for this contribution! Can you include some benchmark results (or a snippet of a profile) to get a sense of the speedup?

File: vllm/v1/attention/backends/mla/common.py (hunk 1 above)
Comment (IwakuraRein) [reply vllm/v1/attention/backends/mla/common.py:2049]: We have done some kernel breakdowns on the GB200 and Flashinfer MLA. It shows that the `CatArrayBatchedCopy_contig` in the flashinfer mla has significant overhead (even longer than FMHA). 

Using `torch.empty` and `copy_` seems to play well with torch.compile. After the change, the `CatArrayBatchedCopy_contig` disappears, and no new kernel shows up. Likely being fused into other operations.

Haven't tried doing `torch.cat` before `scaled_fp8_quant`. Will do more experiments to find out.

Comment (IwakuraRein) [reply vllm/v1/attention/backends/mla/common.py:2049]: In our experiments, we use DEP4, nvfp4 checkpoint, and 1024 tokens per each rank in the decode phase. The projected speedup is 5% in the decode phase.

Comment (IwakuraRein) [reply vllm/v1/attention/backends/mla/common.py:2049]: @MatthewBonanni Just tested `decode_q = torch.cat([decode_ql_nope, decode_q_pe], dim=-1)` and `CatArrayBatchedCopy` still occurs.

Comment (pavanimajety) [reply vllm/v1/attention/backends/mla/common.py:2049]: +1 for reusing `decode_q` or `del decode_q0` after `scaled_fp8_quant`

**Review by pavanimajety — APPROVED** [review]
LGTM, minor comments
